curator-client/src/main/java/org/apache/curator/utils/ZKPaths.java

```diff
@@ -333,26 +341,49 @@ public static void mkdirs(ZooKeeper zookeeper, String path, boolean makeLastNode
             String subPath = path.substring(0, pos);
             if ( zookeeper.exists(subPath, false) == null )
             {
-                try
+                List<ACL> acl = null;
+                if ( aclProvider != null )
                 {
-                    List<ACL> acl = null;
-                    if ( aclProvider != null )
-                    {
-                        acl = aclProvider.getAclForPath(subPath);
-                        if ( acl == null )
-                        {
-                            acl = aclProvider.getDefaultAcl();
-                        }
-                    }
+                    acl = aclProvider.getAclForPath(subPath);
                     if ( acl == null )
                     {
-                        acl = ZooDefs.Ids.OPEN_ACL_UNSAFE;
+                        acl = aclProvider.getDefaultAcl();
                     }
-                    zookeeper.create(subPath, new byte[0], acl, getCreateMode(asContainers));
                 }
-                catch ( KeeperException.NodeExistsException e )
+                if ( acl == null )
+                {
+                    acl = ZooDefs.Ids.OPEN_ACL_UNSAFE;
+                }
+
+                CompletableFuture<KeeperException> f = new CompletableFuture<>();
+
+                zookeeper.create(subPath, new byte[0], acl, getCreateMode(asContainers), new AsyncCallback.Create2Callback()
```

Comment:
The callback-completable-future-mixed style is not quite elegant, but this is the best solution I find now.